### PR TITLE
DLPX-64416 Linux upgrade failure, mgmt stack did not restart after orderly shutdown, upgrade incomplete

### PR DESCRIPTION
**Pull request by prakashsurya**

### CODE REVIEW

**Comment by prakashsurya** [comment]
`git-ab-pre-push` is [here](http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/appliance-build-orchestrator-pre-push/1313/)

**Comment by prakashsurya** [comment]
cc: @grwilson @pzakha

**Comment by prakashsurya** [comment]
Please see [the bug](https://jira.delphix.com/browse/DLPX-64416), and more specifically [my comment](https://jira.delphix.com/browse/DLPX-64416?focusedCommentId=496023&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-496023) for more details on why I think this change is needed.

**Comment by pzakha** [comment]
Can you outline the differences in service files with the `systemd` option removed from `dh`? I believe what you wanted to remove was specifically the `systemctl start/stop` instructions from `postinst/postrm` scripts. What about `systemctl enable/disable`? What else does this affect?

**Comment by prakashsurya** [comment]
This will remove all the automatically added logic. Anything that's specifically added by code in this repository will remain, but all the automatically added logic (including enable/disable, start/stop, etc.) will be removed.

IMO, we should do any enable/disable, and start/stop, that we may need explicitly in one of our delphix packages; e.g. delphix-platform.

I can provide a diff for all the scripts/packages if you'd like to see exactly what's changing; let me know if that'd be useful.

**Comment by pzakha** [comment]
I think it may be useful to include that diff somewhere for tracking purposes.

We will need to implement the changes for enable/disable first and test them along with this change, otherwise this change will break a bunch of existing logic (not-in-place upgrade, migration, reboot, etc.).

**Comment by prakashsurya** [comment]
> We will need to implement the changes for enable/disable first and test them along with this change

Do you have a preference between us adding explicit enable/disable via delphix-platform, or adding scripts to this repository/package? I think either way would work.. and thinking a little bit more, it might be better to do that in this package.

**Comment by pzakha** [comment]
Yeah its probably best to have it enabled here. Latest changes LGTM, pending more testing.

**Comment by prakashsurya** [comment]
Do we rely on the `zfs-zed` service? As is, I think that'll remain disabled with this change, while it's currently enabled on our VMs.

**Comment by pzakha** [comment]
@prakashsurya Yes, we should definitely enable `zfs-zed`.

**Comment by pzakha** [comment]
Can you post all of the postinst postrm files that were autogenerated? I think that would help us.

File: debian/zfs-zed.prerm [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash -eux
+
+case $1 in
+upgrade | remove)
+	systemctl disable zfs-zed.prerm
```
Comment (pzakha) [inline debian/zfs-zed.prerm]: typo

Comment (prakashsurya) [reply debian/zfs-zed.prerm]: thanks

**Comment by prakashsurya** [comment]
> Can you post all of the postinst postrm files that were autogenerated?

Sure. I'll show the contents of `dpkg -e` of the ZFS package before and after this change.

**Comment by prakashsurya** [comment]
Here's a `diff` of the files output by `dpkg -e` for the ZFS packages, before and after this change:
```
$ diff before.txt after.txt 
47,52c47,52
< zfs-modules-4.18.0-1018-azure/DEBIAN/conffiles
< zfs-modules-4.18.0-1018-azure/DEBIAN/control
< zfs-modules-4.18.0-1018-azure/DEBIAN/md5sums
< zfs-modules-4.18.0-1018-azure/DEBIAN/postinst
< zfs-modules-4.18.0-1018-azure/DEBIAN/postrm
< zfs-modules-4.18.0-1018-azure/DEBIAN/triggers
---
> zfs-modules-4.18.0-1019-azure/DEBIAN/conffiles
> zfs-modules-4.18.0-1019-azure/DEBIAN/control
> zfs-modules-4.18.0-1019-azure/DEBIAN/md5sums
> zfs-modules-4.18.0-1019-azure/DEBIAN/postinst
> zfs-modules-4.18.0-1019-azure/DEBIAN/postrm
> zfs-modules-4.18.0-1019-azure/DEBIAN/triggers
59d58
< zfsutils-linux/DEBIAN/postrm
65d63
< zfs-zed/DEBIAN/postrm
```
Here, we can see that we're removing the `postrm` file for both the `zfsutils-linux` and `zfs-zed` packages.

I think it's safe to ignore the files for the azure package, as those differences are an artifact of the kernel version changing between builds, and how I generated the list of files.

This still doesn't show the difference in the files, though, so I'll add another comment after this one to highlight the differences.

Below is information on how I generated the `before.txt` and `after.txt` files.

## Before
#### 1. Generate DEBIAN directory for all ZFS packages
```
$ aws s3 sync --quiet s3://snapshot-de-images/builds/jenkins-ops/devops-gate/master/linux-pkg-build/master/kernel/post-push/116 .
$ ls *zfs*.deb | sed 's/\([^_]*\)_.*/\1/' | xargs -I {} -n 1 bash -c 'mkdir {}; mv *{}*.deb {}; cd {}; dpkg -e *.deb'
$ rm * */*.deb 2>/dev/null
$ find * -maxdepth 1 -mindepth 1
delphix-zfs/DEBIAN
libzfs2linux/DEBIAN
libzfslinux-dev/DEBIAN
zfs-dbg/DEBIAN
zfs-dkms/DEBIAN
zfs-dracut/DEBIAN
zfs-initramfs/DEBIAN
zfs-modules-4.15.0-1033-gcp/DEBIAN
zfs-modules-4.15.0-1035-kvm/DEBIAN
zfs-modules-4.15.0-1040-aws/DEBIAN
zfs-modules-4.15.0-51-generic/DEBIAN
zfs-modules-4.18.0-1018-azure/DEBIAN
zfs-test/DEBIAN
zfsutils-linux/DEBIAN
zfs-zed/DEBIAN
```
#### 2. Generate list of files contained in DEBIAN directories
```
$ find * -type f | sort > ../before.txt
$ head ../before.txt 
delphix-zfs/DEBIAN/control
delphix-zfs/DEBIAN/md5sums
libzfs2linux/DEBIAN/control
libzfs2linux/DEBIAN/md5sums
libzfs2linux/DEBIAN/shlibs
libzfs2linux/DEBIAN/triggers
libzfslinux-dev/DEBIAN/control
libzfslinux-dev/DEBIAN/md5sums
zfs-dbg/DEBIAN/control
zfs-dbg/DEBIAN/md5sums
```

## After
#### 1. Generate DEBIAN directory for all ZFS packages
```
$ aws s3 sync --quiet s3://dev-de-images/builds/jenkins-selfservice/devops-gate/master/linux-pkg-build/master/kernel/pre-push/16 .
$ ls *zfs*.deb | sed 's/\([^_]*\)_.*/\1/' | xargs -I {} -n 1 bash -c 'mkdir {}; mv *{}*.deb {}; cd {}; dpkg -e *.deb'
$ rm * */*.deb 2>/dev/null
```
#### 2. Generate list of files contained in DEBIAN directories
```
$ find * -type f | sort > ../after.txt
```

**Comment by prakashsurya** [comment]
Using the `zfs-zed` package as the example, here's the differences between that packages `postinst` and `prerm` scripts:

#### prerm
```
$ diff -u {before,after}/zfs-zed/DEBIAN/prerm
--- before/zfs-zed/DEBIAN/prerm 2019-06-05 22:19:23.000000000 +0000
+++ after/zfs-zed/DEBIAN/prerm  2019-06-06 18:00:12.000000000 +0000
@@ -1,7 +1,9 @@
-#!/bin/sh
-set -e
-# Automatically added by dh_systemd_start/11.1.6ubuntu2
-if [ -d /run/systemd/system ]; then
-       deb-systemd-invoke stop 'zfs-zed.service' >/dev/null || true
-fi
-# End automatically added section
+#!/bin/bash -eux
+
+case $1 in
+upgrade | remove)
+       systemctl disable zfs-zed.service
+       ;;
+esac
+
+exit 0
```

#### postinst
```
$ diff -u {before,after}/zfs-zed/DEBIAN/postinst
--- before/zfs-zed/DEBIAN/postinst      2019-06-05 22:19:23.000000000 +0000
+++ after/zfs-zed/DEBIAN/postinst       2019-06-06 18:00:12.000000000 +0000
@@ -1,27 +1,9 @@
-#!/bin/sh
-set -e
-# Automatically added by dh_systemd_enable/11.1.6ubuntu2
-if [ "$1" = "configure" ] || [ "$1" = "abort-upgrade" ] || [ "$1" = "abort-deconfigure" ] || [ "$1" = "abort-remove" ] ; then
-       # This will only remove masks created by d-s-h on package removal.
-       deb-systemd-helper unmask 'zfs-zed.service' >/dev/null || true
+#!/bin/bash -eux
 
-       # was-enabled defaults to true, so new installations run enable.
-       if deb-systemd-helper --quiet was-enabled 'zfs-zed.service'; then
-               # Enables the unit on first installation, creates new
-               # symlinks on upgrades if the unit file has changed.
-               deb-systemd-helper enable 'zfs-zed.service' >/dev/null || true
-       else
-               # Update the statefile to add new symlinks (if any), which need to be
-               # cleaned up on purge. Also remove old symlinks.
-               deb-systemd-helper update-state 'zfs-zed.service' >/dev/null || true
-       fi
-fi
-# End automatically added section
-# Automatically added by dh_systemd_start/11.1.6ubuntu2
-if [ "$1" = "configure" ] || [ "$1" = "abort-upgrade" ] || [ "$1" = "abort-deconfigure" ] || [ "$1" = "abort-remove" ] ; then
-       if [ -d /run/systemd/system ]; then
-               systemctl --system daemon-reload >/dev/null || true
-               deb-systemd-invoke start 'zfs-zed.service' >/dev/null || true
-       fi
-fi
-# End automatically added section
+case $1 in
+configure)
+       systemctl enable zfs-zed.service
+       ;;
+esac
+
+exit 0
```

This is a good example because it's small enough to see the changes clearly, but also show how using `--with=systemd` will automatically manipulate any services that a package installed, based on what operation is being performed on the package.

So, here, we can see the `zfs-zed` would be stopped when the package is removed (including package "upgrade", which consists of package removal, followed by package installation) without this change.

This same service manipulation occurs with the `zfsutils-linux` package, but that package will manipulate `zfs.target`, `zfs-mount.service`, `zfs-import.target`, `zfs-import-cache.service`, etc..

**Comment by prakashsurya** [comment]
Additionally, here's a tarball that contains all of the files that I generated for my comments above: 
[delphix-zfs-52.tar.gz](https://github.com/delphix/zfs/files/3263137/delphix-zfs-52.tar.gz) .. This includes all of the files generated by `dpkg -e` for all of the packages, so folks can look more closely at the differences before and after this change, for each specific package.

**Comment by prakashsurya** [comment]
> Can you post all of the postinst postrm files that were autogenerated?

@pzakha Do my comments above address this? Feel free to look through the files for anything I might have missed; I'd rather take a little time to get this right before landing.

**Comment by pzakha** [comment]
Thanks, I've looked at the tarball and it looks like we've captured the essential logic here.

dh seems to add a bunch of other stuff, but my guess is that most of it does not apply to our use case. In any case we should probably test this with a not-in-place upgrade-full test or migration to make sure we haven't missed anything.

**Comment by prakashsurya** [comment]
I've verified "not-in-place" upgrades work with this change. I'm also running a migration test [here](http://selfservice.jenkins.delphix.com/job/devops-gate/job/master/job/blackbox-chained/283/)

@pzakha is there anything else I should do to verify this?

**Pull request merged**